[Wasm-GC] Introduce i31ref types

takikawa opened the pull request:
#### 2215792e9a8d7e1e51f4fd07b139ae3c0aece440
<pre>
[Wasm-GC] Introduce i31ref types
<a href="https://bugs.webkit.org/show_bug.cgi?id=242261">https://bugs.webkit.org/show_bug.cgi?id=242261</a>

Reviewed by Justin Michaud.

Adds support for i31ref types and i31 operations (i31.new, i31.get_s,
i32.get_u) when experimental GC support is turned on.

As Wasm reference types are represented as encoded JS pointer values,
i31refs are represented with encoded JS integer values with the standard
NumberTag for 32-bit integers.

* JSTests/wasm/gc/i31.js: Added.
(module):
(testI31Type):
(testI31New):
(testI31Get):
(testI31JS):
* JSTests/wasm/wasm.json:
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/llint/WebAssembly32_64.asm:
* Source/JavaScriptCore/llint/WebAssembly64.asm:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::addI31New):
(JSC::Wasm::AirIRGenerator::addI31GetS):
(JSC::Wasm::AirIRGenerator::addI31GetU):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addI31New):
(JSC::Wasm::B3IRGenerator::addI31GetS):
(JSC::Wasm::B3IRGenerator::addI31GetU):
* Source/JavaScriptCore/wasm/WasmExceptionType.h:
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::isI31ref):
(JSC::Wasm::isRefWithTypeIndex):
(JSC::Wasm::isValidHeapTypeKind):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmGlobal.cpp:
(JSC::Wasm::Global::set):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::callInformationForCaller):
(JSC::Wasm::LLIntGenerator::callInformationForCallee):
(JSC::Wasm::LLIntGenerator::addArguments):
(JSC::Wasm::LLIntGenerator::addI31New):
(JSC::Wasm::LLIntGenerator::addI31GetS):
(JSC::Wasm::LLIntGenerator::addI31GetU):
* Source/JavaScriptCore/wasm/WasmParser.h:
(JSC::Wasm::Parser&lt;SuccessType&gt;::parseValueType):
* Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
(JSC::Wasm::marshallJSResult):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyHelpers.h:
(JSC::fromJSValue):
* Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
(JSC::Wasm::wasmToJS):
* Source/JavaScriptCore/wasm/js/WebAssemblyGlobalConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: <a href="https://commits.webkit.org/252688@main">https://commits.webkit.org/252688@main</a>
</pre>

justinmichaud commented on Source/JavaScriptCore/llint/WebAssembly64.asm:
```diff
@@ -1257,3 +1257,31 @@ wasmAtomicCompareExchangeOps(_cmpxchg, Cmpxchg,
+    bqeq t0, ValueNull, .throw
+    returni(ctx, t0)
+
+.throw:
```
The spec says to trap if null, but I am wondering: if i31 refs aren't nullable, then shouldn't downcasts and imports trap before getting here? The only way to get here is by passing in null directly? This seems somewhat unfortunate unless I am missing something

takikawa replied:
> if i31 refs aren't nullable, then shouldn't downcasts and imports trap before getting here?

In general it should be possible for a non-null `i31ref` to get passed in here, they share a null with other reference types. For example the type `i31ref` is shorthand for `(ref null i31)` and includes null.

I think it might be possible in the B3/Air tiers to elide the null check though by inspecting the argument type and seeing if it allows null.

justinmichaud commented on JSTests/wasm/gc/i31.js:
```diff
@@ -0,0 +1,219 @@
+   *   (type (func (param i31ref) (result (ref i31))))
+   * )
+   */
+  new WebAssembly.Instance(module("\x00\x61\x73\x6d\x01\x00\x00\x00\x01\x87\x80\x80\x80\x00\x01\x60\x01\x6a\x01\x6b\x6a"));
```
We should make a script or IDE plugin to compile/decompile these automatically

takikawa replied:
What do you think about using the Wasm GC spec's reference interpreter (bundling the [compiled to JS](https://github.com/WebAssembly/spec/tree/main/interpreter#javascript-library) code) to write these tests?  Then the tests can all be written in text format, except those that test malformed binary format code.

Currently the spec interpreter's compile-to-JS mode appears to be broken, but I'm looking into [a possible fix](https://github.com/WebAssembly/spec/pull/1507).

justinmichaud commented on Source/JavaScriptCore/wasm/WasmExceptionType.h:
```diff
@@ -39,6 +39,8 @@ namespace Wasm {
     macro(OutOfBoundsCallIndirect, "Out of bounds call_indirect"_s) \
     macro(NullTableEntry,  "call_indirect to a null table entry"_s) \
     macro(NullReference,  "call_ref to a null reference"_s) \
+    macro(NullI31GetS,  "i31.get_s to a null reference"_s) \
```
Do we need/want two different exception types?

takikawa replied:
I think it's not necessary, I just put both to be able to differentiate the operation name in errors. But we could use a single case and output `"i31.get_<sx> to a null reference"` or a similar error instead.

takikawa replied:
I've collapsed these two exception types into one in the latest push.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmFormat.h:
```diff
@@ -118,6 +118,12 @@ inline bool isFuncref(Type type)
 
+inline bool isI31ref(Type type)
+{
+    ASSERT(Options::useWebAssemblyTypedFunctionReferences());
```
Is this part of WASM GC or Typed Function References?

takikawa replied:
Whoops, this should check the GC option.

takikawa replied:
Fixed now.

justinmichaud commented on Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
```diff
@@ -66,7 +66,7 @@ void marshallJSResult(CCallHelpers& jit, const TypeDefinition& typeDefinition, c
         }
         default: {
-            if (isFuncref(type) || isExternref(type))
+            if (isFuncref(type) || isExternref(type) || isI31ref(type))
```
So is this supposed to be a number, or an instance of the JS type Webassembly.i31ref? I'm not sure what the current status of the spec is, it seems like the JS api document is out of date?

takikawa replied:
The JS API isn't fully decided yet, but I think it's likely that the semantics will be that I31refs that flow to JS will be interpreted as JS Numbers. For the other direction, JS numbers that are in the i31ref range would be normalized/converted to an i31ref (based on discussion in https://github.com/WebAssembly/gc/issues/76).

For now this PR should let Wasm i31refs get treated as JS values, but the other direction currently should fail and error as the conversion logic hasn't been implemented yet.

justinmichaud commented on Source/JavaScriptCore/wasm/wasm.json:
```diff
@@ -15,6 +15,7 @@
         "externref": { "type": "varint7", "value":  -17, "b3type": "B3::Int64" },
         "ref_null":  { "type": "varint7", "value":  -20, "b3type": "B3::Int64" },
         "ref":       { "type": "varint7", "value":  -21, "b3type": "B3::Int64" },
+        "i31ref":    { "type": "varint7", "value":  -22, "b3type": "B3::Void" },
```
Should the B3 type be an Int64?

takikawa replied:
I had left this as `Void` here to be consistent with, e.g., structs. In particular `TypeKind::I31ref` shouldn't show up as a type on its own (same with structs), because the parser should always convert an i31ref type to `TypeKind::Ref` (or `RefNull`) with an I31 heap type.

justinmichaud reviewed (approved):
LGTM as long as we gate everything by the Wasm GC option, not just typed function references. r=me

takikawa committed:
[Wasm-GC] Introduce i31ref types
https://bugs.webkit.org/show_bug.cgi?id=242261

Reviewed by Justin Michaud.

Adds support for i31ref types and i31 operations (i31.new, i31.get_s,
i32.get_u) when experimental GC support is turned on.

As Wasm reference types are represented as encoded JS pointer values,
i31refs are represented with encoded JS integer values with the standard
NumberTag for 32-bit integers.

* JSTests/wasm/gc/i31.js: Added.
(module):
(testI31Type):
(testI31New):
(testI31Get):
(testI31JS):
* JSTests/wasm/wasm.json:
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/llint/WebAssembly32_64.asm:
* Source/JavaScriptCore/llint/WebAssembly64.asm:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::addI31New):
(JSC::Wasm::AirIRGenerator::addI31GetS):
(JSC::Wasm::AirIRGenerator::addI31GetU):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addI31New):
(JSC::Wasm::B3IRGenerator::addI31GetS):
(JSC::Wasm::B3IRGenerator::addI31GetU):
* Source/JavaScriptCore/wasm/WasmExceptionType.h:
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::isI31ref):
(JSC::Wasm::isRefWithTypeIndex):
(JSC::Wasm::isValidHeapTypeKind):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::parseExpression):
(JSC::Wasm::FunctionParser<Context>::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmGlobal.cpp:
(JSC::Wasm::Global::set):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::callInformationForCaller):
(JSC::Wasm::LLIntGenerator::callInformationForCallee):
(JSC::Wasm::LLIntGenerator::addArguments):
(JSC::Wasm::LLIntGenerator::addI31New):
(JSC::Wasm::LLIntGenerator::addI31GetS):
(JSC::Wasm::LLIntGenerator::addI31GetU):
* Source/JavaScriptCore/wasm/WasmParser.h:
(JSC::Wasm::Parser<SuccessType>::parseValueType):
* Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
(JSC::Wasm::marshallJSResult):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyHelpers.h:
(JSC::fromJSValue):
* Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
(JSC::Wasm::wasmToJS):
* Source/JavaScriptCore/wasm/js/WebAssemblyGlobalConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: https://commits.webkit.org/252688@main
